sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -129,7 +129,7 @@ case class CaseWhen(
     case Seq(dt1, dt2) => dt1.sameType(dt2)
   }
 
-  override def dataType: DataType = branches.head._2.dataType
+  override def dataType: DataType = valueTypes.reduce(TypeCoercion.findTightestCommonType(_, _).get)
```

Comment:
I think we need not use `TypeCoercion.findTightestCommonType`, because we've already checked if the types are same in `checkInputDataTypes`. So, I think it's just ok to consider nullability only?

Reply:
Yeah, the types are ```sameType``` equal (ignoring nullability). I just leveraging ```TypeCoercion.findTightestCommonType``` for merging ```nullable``` and ```containsNull``` flags. 

I can introduce a new method for this purpose, but I'm afraid that the implementation will be very similar to ```TypeCoercion.findTightestCommonType```.

Reply:
Ok, will do that :-)